connectivity: rework sniffer to execute tcpdump in background

smagnani96 opened the pull request:
Please do refer to commit messages for a detailed explanation.
I'd love as many feedback as possible, given this ended up being a not so easy change 😅 

1st and 2nd commit are the major updates. Sometimes while re-reading them they do feel quite hacky, but haven't found a better way to implement that.
3rd commit `connectivity: limit tcpdump capture also for Assert mode` is a **Why not?**, definitively can go in a different PR (or even cancel it if I missed the reason why we shouldn't do that).

Fixes: #38643

smagnani96 committed:
connectivity: rework sniffer to execute tcpdump in background

Prior to this commit, we were running remote sniffers via ExecInPodWithWriters
and streaming stdout/stderr through the connection. We observed race conditions
with underlying libraries (both WebSocket and SPDY) in case one of the
stream is not available yet/anymore. In that case, the remote tcpdump command
that is printing packet/logs in the streams would fail with error code 14
(broken pipe). Such flake was reproducible only in CI. The test consisted of
the following instructions:

1. running an ExecInPodWithWriters connection with stdout/stderr stream for tcpdump
2. running an ExecInPod at the end of the test to retrieve packets count
3. running an ExecInPod at the end of the test to retrieve and print packets

With this commit, we run the remote sniffer in background, without keeping
open connections for streaming data. Tcpdump now logs both stdout/stderr
and potential captured packets to local files in the pod, which are then
retrieved/read by us at the end of each test. The difference is in the
way we start/stop such command: given we do not have anymore an open connection
to close, we need to carefully make sure that we call stop in all the possible
circumstances (e.g., context canceled, timeout hit, errors in between two
`Sniff()`, etc.). To do that, we introduce a close/finalizer function that
is being returned in the costructor, so that the caller can defer its execution.
The `Validate()` is also able to stop the sniffer, as well as the constructor
`Sniff()` itself in case of errors (e.g., context canceled or timeout hit).
We do now perform the following operations:

i.   run an ExecInPod to create the remote tcpdump process in background
ii.  run an ExecInPod to stop the remote process at the end of the test
iii. running an ExecInPod at the end of the test to retrieve packets count
iv.  running an ExecInPod at the end of the test to retrieve and print packets

While it is true that it might overload more the API server given we do need
to ExecInPod once more, there's no relevant stdout/stderr in (i) and (ii)
anymore, hopefully mitigating the tcpdump issue we were facing in CI flakes.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>
Co-developed-by: Maxim Mikityanskiy <maxim@isovalent.com>
Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

smagnani96 committed:
connectivity: limit tcpdump capture also for Assert mode

When running in Sanity mode, we do limit the number of captured packets
to one, just to make sure that we see traffic in the interface where the
sniffer is listening to. In Assert mode, we capture all the packets matching
the filter, and we signal the leak at the end of the test by printing all
of them. I'd suggest we limit also in Assert mode the amount of packets
captured. Historically, I've seen captured packets to always repeat
themselves, and I'm not sure it could be very helpful to have, for instance,
more than 1000 leaked packets being logged. In addition, limiting the
amount of captured packets would also help with keeping low memory
footprint, given we do not remove the dump files yet from the pods.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
connectivity: reduce ExecInPod calls when validating a tcpdump sniffer

Prior to this patch series of 3 commits, we were:

1. running an ExecInPodWithWriters connection with stdout/stderr stream for tcpdump
2. running an ExecInPod at the end of the test to retrieve packets count
3. running an ExecInPod at the end of the test to retrieve and print packets

In previous commits, we reworked the way we create and stop such sniffers:
the tcpdump command is now left in background in the pod, and no active
stream connections are present during the related tests. The total amount
of remote commands executions were as follows:

i.   run an ExecInPod to create the remote tcpdump process in background
ii.  run an ExecInPod to stop the remote process at the end of the test
iii. running an ExecInPod at the end of the test to retrieve packets count
iv.  running an ExecInPod at the end of the test to retrieve and print packets

This is clearly inefficient, given we moved to background tcpdump in pods
to avoid error 14 (broken pipe) due to race conditions while setting up
the stream (the flake happened with both WebSockets and SPDY connections),
observed especially under heavy load of the API server.

This commit optimizes such behavior, unifying (ii) - (iii) - (iv).
Upon successfully stopping the remote sniffer, the stop command executed
in (ii) will additionally print to stdout (iii) and (iv) without having
to re-execute ExecInPod. If the sniffer has been successfully stopped,
this command will not fail, even if facing errors while printing out
the packet count (iii) or the packet themselves (iv). This is aligned to
what we're already doing right now, but with less connections.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

giorio94 commented on cilium-cli/connectivity/sniff/sniffer.go:
```diff
@@ -25,17 +25,105 @@ const (
+	done
+
+	# Process is running but did not receive confirmation, kill and exit.
+	kill -2 $pid 2>/dev/null
```
Nit. Any reason for using `SIGINT` vs `SIGKILL`?

gentoo-root replied:
Good point, I'd suggest SIGKILL too, because tcpdump returns exit code 0 when killed with SIGINT, and we don't want to return 0 (success) in this case.

giorio94 commented on cilium-cli/connectivity/sniff/sniffer.go:
```diff
@@ -25,17 +25,105 @@ const (
+	// (tcpdump: Unable to write output: Broken pipe)
+	//
+	// Runs silently: succeeds if tcpdump is active ("listening on <iface>"),
+	// else fails with its error code. Includes `timeout 30` to prevent orphaned
```
Nit. The reference to `timeout 30` looks incorrect.

gentoo-root replied:
Yeah, thanks, looks like it's 60, according to the calculation above.

giorio94 commented on cilium-cli/connectivity/sniff/sniffer.go:
```diff
@@ -25,17 +25,105 @@ const (
+			exit 0
+		fi
+		sleep 1
+	done
```
Any reason for not calling `wait` as for the start case, and let the command context handle the timeout?

gentoo-root replied:
`wait` doesn't have a built-in timeout. You can't run `timeout 30 wait $pid` either, because `wait` is a built-in bash command, but `timeout` is an external one (an executable).

giorio94 replied:
Yeah, my point is that we could theoretically just block forever in this script, and let the caller (i.e., `ExecInPod`) handle the timeout via its context. But the current approach works as well.

giorio94 commented on cilium-cli/connectivity/sniff/sniffer.go:
```diff
@@ -25,17 +25,105 @@ const (
+	sniffConnectionTimeout = sniffScriptTimeout + 5*time.Second
+	// Max remote sniffer runtime to prevent lingering processes in the pod.
+	// NOTE: too low may kill tcpdump while test is running.
+	sniffKillTimeout = sniffConnectionTimeout * 4
```
This is problematic, as there are tests that may require a long-running tcpdump capture, e.g., to verify that certain packets are not seen during the entire connectivity suite execution. I general I don't think it is possible to reliably define a maximum duration of the capture.

gentoo-root replied:
But we do have timeouts on all our tests in the CI (big ones, like 1 hour). If we already have expectations about the duration of the entire test, can we set the tcpdump timeout to that value?

Also, could you please point me at those tests (or how to find them) that run tcpdump for extended periods? We could set different timeouts for short and long tests.

As we run tcpdump in the background, there is no reliable way to kill it after the test finishes, that's why we need those timeouts as a robustness measure.

gentoo-root replied:
> to verify that certain packets are not seen during the entire connectivity suite execution

One that comes to my mind is the one that looks for unencrypted leaks, but it's based on bpftrace, not tcpdump.

pchaigno replied:
> One that comes to my mind is the one that looks for unencrypted leaks, but it's based on bpftrace, not tcpdump.

The current timeout is at 1 min. Won't that be way too low for the tests you mention?

gentoo-root replied:
The CI seems to pass with this timeout. I'd keep it small wherever it works, because some tests spawn multiple sniffers. If we find some long-running test, we can put a custom timeout for that test.

giorio94 commented on cilium-cli/connectivity/sniff/sniffer.go:
```diff
@@ -158,32 +234,40 @@ func (sniffer *Sniffer) Validate(ctx context.Context, a *check.Action) {
+		sniffer.stopCmd = append([]string{"sh", "-c"}, buf.String())
+
+		// Context with timeout for stopping tcpdump.
+		// NOTE: Context is not passed anymore to to this function to avoid premature cancellation.
```
Nit. `s/to to/to/`

giorio94 commented on cilium-cli/connectivity/sniff/sniffer.go:
```diff
@@ -73,13 +73,22 @@ const (
+	# Redirect stderr to /dev/null, as tcpdump logs to stderr.
+	report() {
+		tcpdump -r {{ .DumpPath }} --count 2>/dev/null || true
+		tcpdump -r {{ .DumpPath }}         2>/dev/null || true
```
Nit. Could be worth adding `-n` to avoid IP to name conversions (I know that it was already missing before).

giorio94 commented on cilium-cli/connectivity/sniff/sniffer.go:
```diff
@@ -25,17 +25,105 @@ const (
+	trap '' TERM INT HUP
+
+	# Exit with success code in case tcpdump is not running.
+	if [ ! -f "{{ .PidPath }}" ]; then
```
I assume this should be an error, as nothing should have deleted the PID file.

giorio94 commented on cilium-cli/connectivity/sniff/sniffer.go:
```diff
@@ -197,38 +208,25 @@ func Sniff(ctx context.Context, name string, target *check.Pod,
-			a.Fatalf("Failed to retrieve tcpdump output on %s (%s): %s", sniffer.target.String(), sniffer.target.NodeName(), err)
-		}
-		a.Infof("Captured packets:\n%s", out.String())
+		a.Infof("Captured packets:\n%s", sniffer.out.String())
```
Nit. Should probably trim the first line from the output, that contains the count of the packets.

gentoo-root replied:
I think it's alright as is. `count` has been read with `ReadString()`, which has advanced the position in the buffer, so `String()` will return the remaining stuff (the packets).

giorio94 replied:
Ah, right, I had missed this was acting against a buffer, which then gets consumed.

giorio94 commented on cilium-cli/connectivity/sniff/sniffer.go:
```diff
@@ -197,38 +208,25 @@ func Sniff(ctx context.Context, name string, target *check.Pod,
-	// will return an error if any char is written to stderr. Anyway, the count
-	// is written to stdout.
-	cmd := []string{"/bin/sh", "-c", fmt.Sprintf("tcpdump -r %s --count 2>/dev/null", sniffer.dumpPath)}
-	count, err := sniffer.target.K8sClient.ExecInPod(ctx, sniffer.target.Pod.Namespace, sniffer.target.Pod.Name, "", cmd)
```
It is curious that this one never failed, while the issue was always with the startup of tcpdump. So it is probably not strictly a stdout/stderr issue, but more a problem with some tcpdump expectations there.

gentoo-root replied:
A hypothesis, after looking at a log with my [debug prints](https://github.com/cilium/cilium/pull/40692):

```
2025-07-25T20:56:56.3826162Z XXX: stdout err EOF, sniffer 0xc00091c200
2025-07-25T20:56:56.4852893Z XXX: tcpdump is listening, sniffer 0xc00091c200
2025-07-25T20:56:56.4853573Z XXX: leaving Sniff, sniffer 0xc00091c200
2025-07-25T20:56:56.5962425Z XXX: stopAndWait, sniffer 0xc00091c200
2025-07-25T20:56:56.5963731Z XXX: sniffer.cancel() 0xc00091c200
2025-07-25T20:56:56.6225101Z 
2025-07-25T20:56:56.6228007Z XXX: sniffer 0xc00091c200 exited with err command failed (pod=cilium-test-3/host-netns-2stxb, container=): command terminated with exit code 14
2025-07-25T20:56:56.6232855Z XXX: sniffer.exited 0xc00091c200 after cancel, err command failed (pod=cilium-test-3/host-netns-2stxb, container=): command terminated with exit code 14
2025-07-25T20:56:56.6234036Z XXX: leaving stopAndWait, sniffer 0xc00091c200
```

So, `sniffer.cancel()` canceled `cmdctx`, and **maybe** it first closes the pipes, then kills the process. If tcpdump wrote something in the meantime, it would fail and exit with code 14 by itself, before it's killed.

That also explains why we never fail with error 14 when reading the pcap.

giorio94 reviewed (commented):
Nice! A few drive-by comments.

gentoo-root commented on cilium-cli/connectivity/sniff/sniffer.go:
```diff
@@ -25,17 +25,105 @@ const (
+		sleep 1
+	done
+
+	# Process still exists, exit with error.
```
We can send a SIGKILL at this point, just in case.

gentoo-root reviewed (commented):
Simone is on vacation, I'm taking it over for the while.

gentoo-root commented on cilium-cli/connectivity/sniff/sniffer.go:
```diff
@@ -25,17 +25,116 @@ const (
+	pid=$(cat {{ .PidPath }} )
+
+	# Kill the process if it is still running.
+	if kill -0 $pid 2>/dev/null; then
```
I think we should error out if it's no longer running. If @giorio94 is [right](https://github.com/cilium/cilium/pull/40487#discussion_r2231523844), this will allow us to detect long-running tcpdump tests that would kill tcpdump on timeout, before the test finishes.

gentoo-root committed:
connectivity: Pass -n to tcpdump

Avoid reverse DNS resolution of IP addresses.

Suggested-by: Marco Iorio <marco.iorio@isovalent.com>
Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

pchaigno commented on cilium-cli/connectivity/sniff/sniffer.go:
```diff
@@ -25,17 +25,111 @@ const (
+	// (tcpdump: Unable to write output: Broken pipe)
+	//
+	// Runs silently: succeeds if tcpdump is active ("listening on <iface>"),
+	// else fails with its error code. Includes `timeout 60` to prevent orphaned
```
Nit in case of repush:
```suggestion
	// else fails with its error code. Includes `timeout` to prevent orphaned
```
given the timeout is not hardcoded here.

pchaigno commented on cilium-cli/connectivity/sniff/sniffer.go:
```diff
@@ -163,12 +163,13 @@ func Sniff(ctx context.Context, name string, target *check.Pod,
+	case ModeSanity:
+		count = 1
+	case ModeAssert:
+		count = 1000
```
IMO, even 100 would be largely enough. I wouldn't delay the PR on this though.

pchaigno reviewed (approved):
LGTM! Nice commit history! :pray:

smagnani96 commented on cilium-cli/connectivity/sniff/sniffer.go:
```diff
@@ -25,17 +25,122 @@ const (
+	# Return an error if tcpdump timed out before capturing any packets.
+	if kill -0 $pid 2>/dev/null; then
+		kill -2 $pid 2>/dev/null
+	elif [ "$(tcpdump -r {{ .DumpPath }} --count 2>/dev/null)" = "0 packets" ]; then
```
@gentoo-root I believe this is incorrect for the ModeAssert where having `0 packets` is good.
Isn't this checked later in Go anyway?

gentoo-root replied:
The point is to error out if tcpdump exited prematurely. If it didn't collect any packets and exited, it means it may have missed some packets, so we better fail the test.

In ModeAssert, however, we expect that tcpdump survives until it's killed on line 99. We don't hit the check on line 100 if everything works properly. Only if tcpdump was running during the entire test, we can be sure that 0 packets is indeed 0 packets, and we return 0 on line 111 in this case. So it's all good.

smagnani96 replied:
Nope, given we now have also `-c 1000` for ModeAssert.
Tcpdump might have already recorded 1000 packets and exited before our check here.
